File: sdks/java/src/main/java/com/candelahq/sdk/CandelaSession.java
package com.candelahq.sdk;

import java.net.http.HttpRequest;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.StringJoiner;
import java.util.regex.Pattern;

/**
 * Candela Enrichment SDK for Java.
 *
 * <p>Zero-dependency utility for propagating tenant and job metadata to Candela
 * AI observability proxies via W3C Baggage headers.
 *
 * <h3>Usage with HttpClient:</h3>
 * <pre>{@code
 * CandelaSession session = new CandelaSession.Builder()
 *     .tenantId("acme-corp")
 *     .jobId("training-v3")
 *     .build();
 *
 * HttpRequest.Builder builder = HttpRequest.newBuilder()
 *     .uri(URI.create("http://localhost:8080/v1/chat/completions"));
 * session.injectHeaders(builder);
 * HttpRequest request = builder.POST(...).build();
 * }</pre>
 *
 * <h3>Usage with OkHttp Interceptor:</h3>
 * <pre>{@code
 * OkHttpClient client = new OkHttpClient.Builder()
 *     .addInterceptor(chain -> {
 *         Request original = chain.request();
 *         Request.Builder builder = original.newBuilder();
 *         session.headers().forEach(builder::header);
 *         return chain.proceed(builder.build());
 *     })
 *     .build();
 * }</pre>
 */
public final class CandelaSession {

    private static final Pattern ID_PATTERN =
            Pattern.compile("^[a-zA-Z0-9\\-._]{1,128}$");

    private final String tenantId;
    private final String jobId;

    private CandelaSession(String tenantId, String jobId) {
        this.tenantId = tenantId;
        this.jobId = jobId;
    }

    /**
     * Validate a tenant or job ID against the allowed pattern.
     *
     * @throws IllegalArgumentException if the ID is invalid
     */
    public static String validateId(String value, String name) {
        if (value == null || !ID_PATTERN.matcher(value).matches()) {
            throw new IllegalArgumentException(
                    "candela: invalid " + name + " \"" + value
                            + "\" — must be 1-128 chars of [a-zA-Z0-9._-]");
        }
        return value;
    }

    /**
     * Return enrichment headers as a Map. A fresh Map is returned on each call.
     */
    public Map<String, String> headers() {
        Map<String, String> h = new LinkedHashMap<>();
        StringJoiner baggage = new StringJoiner(",");

        if (tenantId != null) {
            baggage.add("candela.tenant_id=" + tenantId);
            h.put("X-Candela-Tenant-Id", tenantId);
        }
        if (jobId != null) {
            baggage.add("candela.job_id=" + jobId);
            h.put("X-Candela-Job-Id", jobId);
        }
        if (baggage.length() > 0) {
            h.put("Baggage", baggage.toString());
        }
        return h;
    }

    /**
     * Inject enrichment headers into a {@link HttpRequest.Builder}.
     */
    public void injectHeaders(HttpRequest.Builder builder) {
        headers().forEach(builder::header);
    }

    /** Builder for {@link CandelaSession}. */
    public static final class Builder {
        private String tenantId;
        private String jobId;

        public Builder tenantId(String tenantId) {
            this.tenantId = validateId(tenantId, "tenant_id");
            return this;
        }

        public Builder jobId(String jobId) {
            this.jobId = validateId(jobId, "job_id");
            return this;
        }

        public CandelaSession build() {
            return new CandelaSession(tenantId, jobId);
        }
    }
}

File: sdks/kotlin/src/main/kotlin/com/candelahq/sdk/CandelaSession.kt
package com.candelahq.sdk

/**
 * Candela Enrichment SDK for Kotlin/JVM.
 *
 * Zero-dependency middleware for propagating tenant and job metadata
 * to Candela AI observability proxies via W3C Baggage headers.
 *
 * ```kotlin
 * val session = candelaSession {
 *     tenantId = "acme-corp"
 *     jobId = "training-v3"
 * }
 *
 * // With OkHttp
 * val client = OkHttpClient.Builder()
 *     .addInterceptor(session.okHttpInterceptor())
 *     .build()
 *
 * // With any HTTP client
 * val headers = session.headers()
 * ```
 */

private val ID_PATTERN = Regex("^[a-zA-Z0-9\\-._]{1,128}$")

/**
 * Validate a tenant or job ID against the allowed pattern.
 * @throws IllegalArgumentException if the ID is invalid.
 */
@JvmOverloads
fun validateId(value: String, name: String = "id"): String {
    require(ID_PATTERN.matches(value)) {
        "candela: invalid $name \"$value\" — must be 1-128 chars of [a-zA-Z0-9._-]"
    }
    return value
}

/**
 * Reusable session that generates enrichment headers for all requests.
 *
 * Construct via DSL:
 * ```kotlin
 * val session = candelaSession {
 *     tenantId = "acme"
 *     jobId = "run-42"
 * }
 * ```
 *
 * Or from Java:
 * ```java
 * CandelaSession session = CandelaSession.create("acme", "run-42");
 * ```
 */
class CandelaSession private constructor(
    val tenantId: String?,
    val jobId: String?,
) {
    init {
        tenantId?.let { validateId(it, "tenant_id") }
        jobId?.let { validateId(it, "job_id") }
    }

    /** Return a fresh map of enrichment headers. */
    fun headers(): Map<String, String> = buildMap {
        val parts = mutableListOf<String>()

        tenantId?.let {
            parts += "candela.tenant_id=$it"
            put("X-Candela-Tenant-Id", it)
        }

        jobId?.let {
            parts += "candela.job_id=$it"
            put("X-Candela-Job-Id", it)
        }

        if (parts.isNotEmpty()) {
            put("Baggage", parts.joinToString(","))
        }
    }

    /**
     * Inject enrichment headers into an existing mutable map.
     * Preserves existing Baggage entries.
     */
    fun injectHeaders(headers: MutableMap<String, String>): MutableMap<String, String> {
        val ours = headers()
        for ((k, v) in ours) {
            if (k == "Baggage") {
                val existing = headers[k]
                headers[k] = if (existing != null) "$existing,$v" else v
            } else {
                headers[k] = v
            }
        }
        return headers
    }

    /** DSL builder for [CandelaSession]. */
    class Builder {
        var tenantId: String? = null
        var jobId: String? = null

        fun build(): CandelaSession = CandelaSession(tenantId, jobId)
    }

    companion object {
        /** Create a session from Java. */
        @JvmStatic
        @JvmOverloads
        fun create(tenantId: String? = null, jobId: String? = null): CandelaSession =
            CandelaSession(tenantId, jobId)
    }
}

/** DSL function to create a [CandelaSession]. */
fun candelaSession(block: CandelaSession.Builder.() -> Unit): CandelaSession =
    CandelaSession.Builder().apply(block).build()
